net/tcp/tcp_timer.c

```diff
@@ -74,10 +76,192 @@
 
 #define ACK_DELAY (1)
 
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: tcp_get_timeout
+ *
+ * Description:
+ *   Gets the time of the next timeout
+ *
+ * Input Parameters:
+ *   conn - The TCP "connection" to poll for TX data
+ *
+ * Returned Value:
+ *   int - The time required for the next expiry (units: half-seconds)
+ *
+ * Assumptions:
+ *   conn is not NULL.
+ *   The connection (conn) is bound to the polling device (dev).
+ *
+ ****************************************************************************/
+
+static int tcp_get_timeout(FAR struct tcp_conn_s *conn)
+{
+  int timeout = conn->timer;
+
+#ifdef CONFIG_NET_TCP_KEEPALIVE
+  if (timeout == 0)
+    {
+      timeout = conn->keeptimer;
+    }
+  else if (conn->keeptimer > 0 && timeout > conn->keeptimer)
+    {
+      timeout = conn->keeptimer;
+    }
+#endif
+
+  return timeout;
+}
+
+/****************************************************************************
+ * Name: tcp_timer_expiry
+ *
+ * Description:
+ *   Handle a TCP timer expiration for the provided TCP connection
+ *   Restart a TCP timer if need to
+ *
+ * Input Parameters:
+ *   arg - The TCP "connection" to poll for TX data
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   arg is not NULL.
+ *   The connection (arg) is bound to the polling device (dev).
+ *
+ ****************************************************************************/
+
+static void tcp_timer_expiry(FAR void *arg)
+{
+  FAR struct tcp_conn_s *conn = arg;
+
+  conn->timeout = true;
+  conn->dev->d_txavail(conn->dev);
+}
+
+/****************************************************************************
+ * Name: tcp_update_timer
+ *
+ * Description:
+ *   Update the TCP timer for the provided TCP connection,
+ *   The timeout is accurate
+ *
+ * Input Parameters:
+ *   conn - The TCP "connection" to poll for TX data
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   conn is not NULL.
+ *   The connection (conn) is bound to the polling device (dev).
+ *
+ ****************************************************************************/
+
+static void tcp_update_timer(FAR struct tcp_conn_s *conn)
+{
+  int ticks = HSEC2TICK(tcp_get_timeout(conn));
+
+  if (ticks > 0)
+    {
+      sclock_t left = work_timeleft(&conn->work);
+      if (left <= 0 || llabs(left - ticks) >= HSEC2TICK(1))
```

Comment:
Could you please describe more about `llabs(left - ticks) >= HSEC2TICK(1)`. It is not obvious for me.
Are we trying to adjust work queue timeout in case current timeout is less than `tcp_get_timeout()`?

Reply:
All timer in tcp use half second as the unit:
https://github.com/apache/incubator-nuttx/blob/master/net/tcp/tcp.h#L195
https://github.com/apache/incubator-nuttx/blob/master/net/tcp/tcp.h#L287-L292
so we can save the unnecessary work operation if the difference between the left time of work and the timeout value is smaller than the half second.

Reply:
Then maybe we can avoid `llabs(left - ticks)` usage, because it is not obvious.
Maybe we can
```
int timeout = tcp_get_timeout(conn);

if (timeout > 0)
  {
    sclock_t left = work_timeleft(&conn->work);
    if (left <= 0 && TICK2HSEC(left) > timeout)
      {
        work_queue(LPWORK, &conn->work, tcp_timer_expiry,
                   conn, HSEC2TICK(timeout));
      }
  }
else
  {
    work_cancel(LPWORK, &conn->work);
  }
```
or anything similar.
Maybe I understand logic in a wrong way. I do not understand `llabs(left - ticks) >= HSEC2TICK(1)`. So if `HSEC2TICK(tcp_get_timeout(conn))` is `5000` and left is `3000`, so condition will `llabs(3000 - 5000)` -> `2000` so  we will start `work_queue` with `5000` timeout.
So my question is when do we need to start `work_queue`? If it is possible to write as a bullet list?